osquery/tables/system/windows/wmi.cpp

```diff
@@ -0,0 +1,253 @@
+*
+* @returns None.
+*/
+static std::wstring_convert<
+    std::codecvt_utf8_utf16<wchar_t, 0x10ffff, std::little_endian>>
+    converter;
+
+/**
+* @brief Windows helper function for converting narrow strings to wide
+*
+* @returns A wide string, constructed from a narrow string
+*/
+static std::wstring string_to_wstring(const std::string& src) {
+  std::wstring utf16le_str = converter.from_bytes(src);
+  return utf16le_str;
+}
+
+/**
+* @brief Windows helper function for converting wide strings to narrow
+*
+* @returns A narrow string, constructed from a wide string
+*/
+static std::string wstring_to_string(const wchar_t* src) {
+  if (src == nullptr) {
+    return std::string("");
+  }
+
+  std::string utf8_str = converter.to_bytes(src);
+  return utf8_str;
+}
+
+/**
+* @brief Windows WMI Helper function to print the type associated with results
+*
+* @returns A string created from a BSTR
+*/
+static std::string BSTR_to_string(const BSTR src) {
+  return wstring_to_string(static_cast<const wchar_t*>(src));
+}
+
+WmiResultItem::WmiResultItem(WmiResultItem&& src) {
+  result_ = nullptr;
+  std::swap(result_, src.result_);
+}
+
+WmiResultItem::~WmiResultItem() {
+  if (result_ != nullptr) {
+    result_->Release();
+    result_ = nullptr;
+  }
+}
+
+void WmiResultItem::PrintType(const std::string& name) const {
+  std::wstring property_name = string_to_wstring(name);
+  VARIANT value;
+  HRESULT hr = result_->Get(property_name.c_str(), 0, &value, nullptr, nullptr);
+  if (hr != S_OK) {
+    std::cerr << "Failed: " << name << "\n";
+  } else {
+    std::cout << "Name=" << name << ", Type=" << value.vt << "\n";
+    if (value.vt == VT_I4) {
+      std::cout << "  Value=" << value.lVal << "\n";
+    } else if (value.vt == VT_BSTR) {
+      std::wcout << "  Value=" << value.bstrVal << "\n";
+    }
+  }
+}
+
+Status WmiResultItem::GetBool(const std::string& name, bool& ret) const {
+  std::wstring property_name = string_to_wstring(name);
+  VARIANT value;
+  HRESULT hr = result_->Get(property_name.c_str(), 0, &value, nullptr, nullptr);
+
+  if (hr != S_OK) {
+    return Status(-1, "Error retrieving data from WMI query.");
+  }
+  if (value.vt != VT_BOOL) {
+    return Status(-1, "Invalid data type returned.");
+  }
+  ret = value.boolVal;
+  return Status(0);
+}
+
+Status WmiResultItem::GetLong(const std::string& name, long& ret) const {
+  std::wstring property_name = string_to_wstring(name);
+  VARIANT value;
+  HRESULT hr = result_->Get(property_name.c_str(), 0, &value, nullptr, nullptr);
+  if (hr != S_OK) {
+    return Status(-1, "Error retrieving data from WMI query.");
+  }
+  if (value.vt != VT_I4) {
+    return Status(-1, "Invalid data type returned.");
+  }
+  ret = value.lVal;
+  return Status(0);
+}
+
+Status WmiResultItem::GetString(const std::string& name,
+                                std::string& ret) const {
+  std::wstring property_name = string_to_wstring(name);
+  VARIANT value;
```

Comment:
Is there any cleanup that must be done for `VARIANT`s? Such as [`VariantClear`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms221165(v=vs.85).aspx)?